File: sort.js
//VARIABLES
//Get from user
var new_array = document.getElementById("new_array_btn");
var sort_btn = document.getElementById("sort_btn");
var bars_container = document.getElementById("bars_container");
var select_algo = document.getElementById("algo");
var speed = document.getElementById("speed");
var slider = document.getElementById("slider");

//variables for user values
var minRange = 1;
var maxRange = slider.value;
var numOfBars = slider.value;
var heightFactor = 4;
var speedFactor = 100;
var unsorted_array = new Array(numOfBars);

//get the algorithm selected by the user
var algo = "";
sort_btn.addEventListener("click", function () {
  if (algo === "selection") {
    selectionSort(unsorted_array);
  } else if (algo === "insertion") {
    insertionSort(unsorted_array);
  } else if (algo === "heap") {
    heapSort(unsorted_array);
  } else if (algo === "bubble") {
    bubbleSort(unsorted_array);
  } else if (algo === "merge") {
      mergeSort(unsorted_array);
  } else if (algo === "quick") {
    quickSort(unsorted_array, 0, unsorted_array.length - 1);
  } else {
    bubbleSort(unsorted_array);
  }
})

//EVENT LISTENERS:
//take slider input of number of items in array and maximum height of a bar
slider.addEventListener("input", function () {
  numOfBars = slider.value;
  maxRange = slider.value;
  bars_container.innerHTML = "";
  unsorted_array = createRandomArray();
  renderBars(unsorted_array);
});

//take in speed asked by user
speed.addEventListener("change", (e) => {
  speedFactor = parseInt(e.target.value);
});

//take in the algorithm selected, "change" because selected by user
select_algo.addEventListener("change", function () {
  algo = select_algo.value;
});

//load in the array as bars
document.addEventListener("DOMContentLoaded", function () {
  unsorted_array = createRandomArray();
  renderBars(unsorted_array);
});

//generate a new array
new_array.addEventListener("click", function () {
  unsorted_array = createRandomArray();
  bars_container.innerHTML = "";
  renderBars(unsorted_array);
});


//FUNCTIONS TO AID EVENTS AND ALGORITHMS:
//to add a delay
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

//creates the set of bars on the screen
function renderBars(array) {
  //for each bar
  for (let i = 0; i < numOfBars; i++) {
    let bar = document.createElement("div");
    //add a CSS component to bars for colouring and height changing
    bar.classList.add("bar");
    bar.style.height = array[i] * heightFactor + "px";
    //add this 
    bars_container.appendChild(bar);
  }
}

//To fill colour in bars after they go green/red during swaps
function fillBars(exclude1, exclude2){
  let bars = document.getElementsByClassName("bar");
  for (let i = 0; i < bars.length; i++) {
    if(i != exclude1 || i!= exclude2){
      bars[i].style.backgroundColor = "black";
    }
  }
}

//random array to be made into bars
function createRandomArray() {
  let array = new Array(numOfBars);
  for (let i = 0; i < numOfBars; i++) {
    array[i] = randomNum(minRange, maxRange);
  }

  return array;
}

function randomNum(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

//SORTING ALGORITHMS
async function selectionSort(array){
    let bars = document.getElementsByClassName("bar");
    //int placeHolder;
    var minElement;
    var count = 0;
  
    while(count+1<array.length){
        minElement = array[count+1];
        //find min element
        var minElementIndex = count+1;
        for(var i = count+1; i<array.length; i++){
            if(array[i] < minElement){
                minElement = array[i];
                minElementIndex = i;
            }
        }
        //swap place holder and smallest element if it smallest element is smaller
        var placeHolder = array[count];
        if(placeHolder > minElement){
          fillBars(-1, -1);
          array[count] = minElement;
          array[minElementIndex] = placeHolder;
          //give the current bars coloour and update their height on the screen
          bars[count].style.height = array[count] * heightFactor + "px";
          bars[count].style.backgroundColor = "lightgreen";
          bars[minElementIndex].style.height = array[minElementIndex] * heightFactor + "px";
          bars[minElementIndex].style.backgroundColor = "lightgreen";
          //delay
          await sleep(speedFactor);
        }
        
        count++;
    }
    fillBars(-1, -1);
    return array;
  }
  
async function insertionSort(array){
    let bars = document.getElementsByClassName("bar");
    var firstIndex;
    var secondIndex = 1;
    var tempValue;
    //move through the array until the end
    while(secondIndex < array.length){
      tempValue = array[secondIndex];
      firstIndex = secondIndex -1;
      //only the firstIndex is being updated so dont use secondIndex for any
      //keep swapping back until in order by moving the previous one ahead
      while(firstIndex >= 0 && tempValue < array[firstIndex]){
          array[firstIndex+1] = array[firstIndex];
          
          //give the current bars coloour and update their height on the screen
          bars[firstIndex+1].style.height = array[firstIndex+1] * heightFactor + "px";
          bars[firstIndex + 1].style.backgroundColor = "lightgreen";
          //delay
          await sleep(speedFactor);
  
          //update colour of all bars, (lightgreen to black)
          for (let k = 0; k < bars.length; k++) {
            if (k != firstIndex + 1) {
              bars[k].style.backgroundColor = "black";
            }
          }
        firstIndex--;
  
      }
      //since we kept the original second index now we can update it to the correct place
      array[firstIndex+1] = tempValue;
      bars[firstIndex + 1].style.height = array[firstIndex + 1] * heightFactor + "px";
      secondIndex++;
    }
    //update colour of all bars (essentially just the last lightgreen bar)
    for (let k = 0; k < bars.length; k++) {
      bars[k].style.backgroundColor = "black";
    }
    //done
    return array;
  }

File: sort.java
import java.util.Arrays;
public class sort {
    public static void main(String[] args) {
        int[] array = {5, 3, 8, 2, 1, 7, 6, 4};

        System.out.println("Original Array: " + Arrays.toString(array));

        //CALL METHOD
        //bubbleSort(array);
        //heapSort(array);
        //selectionSort(array);
        //insertionSort(array);
        //mergeSort(array);
        //quickSort(array, 0, array.length-1);
        

        System.out.println("Sorted Array: " + Arrays.toString(array));
    }
    //Algorithms
    
    public static void mergeSort(int[] array) {
        //if empty or sorted
        if (array == null || array.length <= 1) {
            return; 
        }
        int[] temp = new int[array.length];
        mergeSort(array, temp, 0, array.length - 1);
    }

    public static void mergeSort(int[] array, int[] temp, int low, int high) {
        if (low < high) {
            int mid = (low + high) / 2;
            //merge sort left and right halfs
            mergeSort(array, temp, low, mid);
            mergeSort(array, temp, mid + 1, high);
            
            //Put the two halfs together:
            
            //make a copy of the array
            for (int i = low; i <= high; i++) {
                temp[i] = array[i];
            }
            //store indices
            int left = low;
            int right = mid + 1;
            int current = low;
            //merge array and temp array
            while (left <= mid && right <= high) {
                if (temp[left] <= temp[right]) {
                    array[current] = temp[left];
                    left++;
                } else {
                    array[current] = temp[right];
                    right++;
                }
                current++;
            }
            //if any elements remain then put then into the final array
            while (left <= mid) {
                array[current] = temp[left];
                left++;
                current++;
            }
            }
    }
    
        
    
    

    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
        //Partition
        //far right is pivot
        int pivot = arr[high];

        //lower element index
        int i = low - 1;

        //iterate through array
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                //Swap if element is less than equal to pivot
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        //Swap
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        int pivotIndex = i + 1;
    
        //Left and right subarrays
        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
        }
    }
    
    
    

    public static void bubbleSort(int[] array){
        int i = 0;
        boolean swapped;
        //before second last element
        while(i < array.length-1){
            swapped = false;
            int j = 0;
            //find max, go until the length-1 - i(number of times we have put the value at the end)
            while(j < array.length-(i+1)){
                if(array[j] > array[j+1]){
                    //swap
                    int temp = array[j];
                    array[j] = array[j+1];
                    array[j+1] = temp;
                    swapped = true;
                }
                j++;
            }
            //no elements were swapped
            if(swapped == false){
                return;
            }
            i++;
        }

    }


    public static void heapSort(int[] array){
        //first convert array into a heap data structure
        //we can do this by calling a heapify function starting from last non leaf node n/2-1
        for(int i = array.length/2 -1; i>=0; i--){
            heapify(array, array.length, i);
        }
        
        int i = array.length-1;
        while(i>0){
            //delete the root node and replace it with the last node in the heap
            //remove new last node, and heapify the elements again
            //continue process until sorted array, then reverse the order
            int temp = array[0];
            array[0] = array[i];
            array[i] = temp;
            heapify(array, i, 0);
            i--;
        }


    }

    public static void heapify(int[] array, int length, int rootIndex){
        int left = 2*rootIndex + 1;
        int right = 2*rootIndex + 2;
        int root = rootIndex;
        //largest with left child, if child is bigger then assign value to largest
        if (left < length && array[left] > array[root]) {
            root = left;
        }
        //largest with right child
        if (right< length && array[right] > array[root]) {
            root = right;
        }
        // If the root is not the largest, swap and recursively heapify the affected subtree
        if (root != rootIndex) {
            int temp = array[rootIndex];
            array[rootIndex] = array[root];
            array[root] = temp;
            heapify(array, length, root);
        }
    }



    public static void selectionSort(int[] array){
        //int placeHolder;
        int minElement;
        int count = 0;

        while(count+1<array.length){
            minElement = array[count+1];
            //find min element
            int minElementIndex = count+1;
            for(int i = count+1; i<array.length; i++){
                if(array[i] < minElement){
                    minElement = array[i];
                    minElementIndex = i;
                }
            }
            //swap place holder and smallest element if it smallest element is smaller
            int placeHolder = array[count];
            if(placeHolder > minElement){
                array[count] = minElement;
                array[minElementIndex] = placeHolder;
            }
            
            count++;
        }
    }


    public static void insertionSort(int[] array){
        //compare first two elements, swap if out of order, new first element is now in a sorted sub-array

        //compare second and third elements, swap if out of order, new second element is now in sorted sub array with first element
        //      only if second is bigger than first

        //compare third and fourth elements, swap if out of order, then if swap compare new third with second (and first if necessary)
        int firstIndex;
        int secondIndex = 1;
        int tempValue;
        //move through array until the end of the array
        while(secondIndex < array.length){
            tempValue = array[secondIndex];
            firstIndex = secondIndex -1;

            //only the firstIndex is being updated so dont use secondIndex for any
            //keep swapping back until in order by moving the previous one ahead
            while(firstIndex >= 0 && tempValue < array[firstIndex]){
                array[firstIndex+1] = array[firstIndex];
                firstIndex--;
            }
            //since we kept the original second index now we can update it to the correct place
            array[firstIndex+1] = tempValue;
            secondIndex++;
        }
        
    
    }
    
}
